remove deprecated feature of passing HostConfig at API container start

mountkin opened the pull request:
- relates to https://github.com/moby/moby/pull/17799
- relates to https://github.com/moby/moby/issues/14777

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**\- What I did**
According to https://docs.docker.com/engine/deprecated/#hostconfig-at-api-container-start passing `HostConfig` to the container start API will be removed in 1.12.
This PR removed the deprecated feature and the related test cases.
**\- How I did it**
Remove the related code.
Raise an error when the user tries to pass HostConfig when calling `POST /containers/xxx/start` API.
**\- How to verify it**
Call the container start API with HostConfig and check the error message.
**\- Description for the changelog**

<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

**\- A picture of a cute animal (not mandatory but encouraged)**

Signed-off-by: Shijiang Wei mountkin@gmail.com

mountkin committed:
remove deprecated feature of passing HostConfig at API container start

Signed-off-by: Shijiang Wei <mountkin@gmail.com>

runcom commented:
LGTM :)

thaJeztah commented:
I don't know how we usually handled API deprecations like this. Looking at this, I wonder if we actually *can* remove this, without disabling API versions 1.something and older.

The reason for a versioned API is of course that it doesn't change, so wondering if we should disable older API versions altogether. Open to suggestions, just thinking out loud here

mountkin commented:
@thaJeztah Could you provide some information or previous discussion about deprecating this feature? 
Indeed removing this code might break some old apps.  I've no idea whether we should keep the current behavior for old API users. Suggestions are welcomed.
Thanks :smiley:

mountkin commented:
Found #17799 and  #14777

thaJeztah commented:
@mountkin yes, I'm trying to think of other examples, i.e., have we done this before? Usually, we keep the old behavior around for older API versions, so perhaps we should change it to be an error for API version 1.24 and above, but keep the old behavior for older versions.


ping @cpuguy83 @calavera wdyt?

cpuguy83 commented:
I don't think we can just remove it, which means we'll have to still handle older API versions, which means the case must still be handled (by switching on API versions) which will probably make the code more complex and not less, sadly.

LK4D4 commented:
I agree with @cpuguy83

thaJeztah commented:
Alright, so we should keep this code around (unfortunately), and make it an error if it's used on API 1.24 or above (instead of just a warning).

@mountkin can you update the PR to do that? Sorry for that, I think we all wished we could cleanup this code

mountkin commented:
I think we can add an HTTP header (for example: `Deprecated-Features-Used: 17799`,  `17799` is the github issue where the deprecation was proposed) to all the HTTP requests to the official registry. 
Then the Docker team can analyse the docker hub HTTP log to determine whether the deprecated feature can be completely removed.
We're already sending kernel version and os version information via the `User-Agent` header.

mountkin commented:
Code updated.
I also moved the related API tests to a separate file `integration-cli/docker_deprecated_api_v124_test.go` and renamed all the test cases to `TestDeprecatedXXX` so that we can run `TESTFLAGS='-check.f DockerSuite.TestDeprecated*' make test-integration-cli` to test the deprecated APIs.

thaJeztah commented:
ping @LK4D4 @cpuguy83 PTAL!

icecrime commented:
Ping @runcom and @calavera too :-)

calavera commented:
:metal: LGTM

calavera commented:
It looks like WinTP5 complains in those new tests, btw.

mountkin commented:
All tests passed. PTAL again. Thanks

vdemeester commented:
LGTM :+1:

vdemeester commented:
Putting it to docs-review, @thaJeztah :stuck_out_tongue_closed_eyes:

thaJeztah commented:
I think we should add a mention in the API changes;
https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v124-api-changes

something like;

```Markdown
* `POST /containers/(id or name)/start` no longer accepts a `HostConfig`
```

mountkin commented:
doc updated

thaJeztah commented:
Thanks! LGTM

ping @vdemeester ptal

vdemeester commented:
LGTM 🐮
